sidecar: add ticket with TLV encoders

guggero opened the pull request:
This is a first preview of the sidecar ticket type and its TLV and string encoders.

> Ticket is a struct holding all the information for establishing/buying a
> sidecar channel. It is meant to be used in a PSBT like manner where each
> participant incrementally adds their information according to their role and
> the current step.

The ticket will replace all new parameters/flags described in https://github.com/guggero/pool/blob/sidecar/docs/sidecar_channels.md.
Instead the ticket will be passed between the participants and enriched with more information in each step.

Roasbeef commented on sidecar/interface.go:
```diff
@@ -0,0 +1,230 @@
+// committing to when offering to buy a channel for the recipient. The sidecar
+// channel flow is initiated by the provider creating a ticket and adding its
+// signed offer to it.
+type Offer struct {
```
Likely don't need in this version, but I imagine we'll also eventually have the following fields:
  * pubkey of the node that the channel may be opened to 
  * type of channel

guggero replied:
I'll keep that in mind. With the extensibility of the TLV stream we should be able to easily add those in a later version.
It got me thinking though... Should the individual fields of the sub structs (`Offer`, `Receiver` and so on) also be encoded as TLV records? If I understood the TLV library correctly you can have records of a stream missing and can add new records. But within one record (for example looking at `DOffer`), all the fields that are read with `tlv.DUint64` still need to be present or you run into an EOF.

Could there be an issue with encoding TLV records inside TLV records or should I go ahead and refactor the sub structs to also use records instead of just fields?

guggero replied:
I went ahead and also used tlv records for the sub structs.
The nice side effect of this is that this also makes it easier to parse the ticket in other languages (for example in JS for a "light" implementation of the sidecar receiver client) as the tlv parsing can just be done recursively.

Roasbeef commented on sidecar/interface.go:
```diff
@@ -0,0 +1,230 @@
+// the current step.
+type Ticket struct {
+	// ID is a pseudorandom identifier of the ticket.
+	ID [8]byte
```
Any reason 8 and not 32?

guggero replied:
Mostly to save some space. This doesn't need to be cryptographically strong IMO, just some unique identifier to tell them apart at the UI level.

Roasbeef replied:
Hmm, so then in this case the issuer is the one responsible for uniqueness?

Roasbeef replied:
I guess in the end, the nonce of the order is the thing that really binds it all.

Roasbeef commented on sidecar/tlv.go:
```diff
@@ -0,0 +1,377 @@
+		))
+	}
+
+	tlvStream, err := tlv.NewStream(records...)
```
I thought the idea was to use _two_ logical streams to allow their contents to be signed independently?

guggero replied:
Discussed this offline. We did mention using two streams initially but during implementation I noticed that the pattern of having a distinct `Digest()` method and then embedding the signature inside the TLV stream itself makes things easier. Mostly for also storing a ticket in the DB as we can just serialize the whole stream and put it into `etcd` or `bbolt`.

Roasbeef commented:
Pondering if we should add a static prefix for the encoded string versions just so it's a bit more identifiable for human users...

halseth commented on sidecar/tlv.go:
```diff
@@ -0,0 +1,377 @@
+)
+
+const (
+	idType        tlv.Type = 1
```
any given reason for using odd numbers here? Since it is a custom type.

guggero replied:
No, I just copied the pattern I found in https://github.com/lightningnetwork/lnd/blob/master/channeldb/invoices.go#L129.
I probably need to read up on the implication of even/odd types. Or does this only apply to the LN gossip wire protocol?

guggero replied:
Ah, I think I get it now. If a type is _unknown_ and is odd, the tlv library will just ignore it. But since we know all our types here, we don't need to use odd types. Fixed.

halseth commented on sidecar/tlv.go:
```diff
@@ -0,0 +1,377 @@
+
+// EPubKeyN is a nil safe Encoder for the btcec.PublicKey type. An error is
+// returned if val is not a **btcec.PublicKey.
+func EPubKeyN(w io.Writer, val interface{}, buf *[8]byte) error {
```
kinda feels like a lot of boiler plate here. Shouldn't most of the logic we need (writing bytes to a TLV stream) already be defined somewhere it can be reused?

guggero replied:
There are pre-defined functions for many things available, yes. But there wasn't one for conditionally serializing a pubkey depending on whether it was nil or not. But I think we should instead use individual TLV records everywhere then the nilness problem would go away.

halseth commented on sidecar/codec.go:
```diff
@@ -0,0 +1,39 @@
+const (
+	// defaultEncodeBufSize is the default buffer size we use when encoding
+	// a ticket.
+	defaultEncodeBufSize = 512
```
Is this really needed?

guggero replied:
No, left over from a previous design. Removed.

halseth reviewed (commented):
Interesting approach!

guggero commented:
> Pondering if we should add a static prefix for the encoded string versions just so it's a bit more identifiable for human users...

Great idea. I did some base64 magic and added a prefix that will spell out "sidecar" when encoded as base64 :grin: 

The other idea I implemented in a first draft was to put everything through gzip to reduce the size. But I then discarded it because that adds another library dependency for anyone trying to implement this in another language.
And I think it doesn't really matter how long the string is. Or do we care about putting it into a tweet for example?

guggero commented:
I also added a checksum to the string encoded version as an additional integrity check.

halseth commented on sidecar/tlv.go:
```diff
@@ -0,0 +1,386 @@
+		return o, err
+	}
+
+	o.Capacity = btcutil.Amount(capacity)
```
probably doesn't matter, but check type was parsed before setting it for consistency?

halseth replied:
never mind, later I see we decode directly into the structs, which is kinda nice way to shorten the code.

halseth commented on sidecar/codec_test.go:
```diff
@@ -0,0 +1,68 @@
+)
+
+var (
+	hardcodedTicket = "sidecarAAQgHBgUEAwIBAAIBYwMBAgp5CwgAAAAAAAADCQwIAA" +
```
😀

Roasbeef replied:
base64 is kinda ugly imo, but hey, it encodes lol

Roasbeef replied:
also we can also change it later if really needed

guggero replied:
Yeah, it's also not optimal because of the special characters (that's why I chose URL safe encoding here). Which encoding did you have in mind? bech32? I chose base64 for two reasons: A bit shorter than bech32 (double the alphabet size) and because PSBT uses the same (trying to convey the idea that this is similar in its concept as a transport format).

halseth reviewed (approved):
This iteration looks very good!

Need to see the bigger picture, but the serilazation code now became less verbose and easy to follow IMO. LGTM ✅

Roasbeef reviewed (approved):
LGTM 🚡
